File: src/main/java/de/tungbeier/paranche/PostService.java
package de.tungbeier.paranche;

import de.tungbeier.paranche.results.Conflict;
import de.tungbeier.paranche.results.Failure;
import de.tungbeier.paranche.results.Found;
import de.tungbeier.paranche.results.NotFound;
import de.tungbeier.paranche.results.RegisterResult;
import de.tungbeier.paranche.results.SearchResult;
import de.tungbeier.paranche.results.Success;
import io.quarkus.logging.Log;
import io.quarkus.panache.common.Parameters;
import io.smallrye.mutiny.Uni;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.hibernate.exception.ConstraintViolationException;

@ApplicationScoped
public class PostService {
    @Inject
    PostRepository postRepository;

    public Uni<RegisterResult> register(Post post) {
        return postRepository
                .persistAndFlush(post)
                .map(p -> (RegisterResult) new Success(p))
                .onFailure().recoverWithItem(error -> {
                    Log.error("Failed to register a post", error);
                    if (error instanceof ConstraintViolationException) {
                        return new Conflict(post, error);
                    } else {
                        return new Failure(post, error);
                    }
                });
    }

    public Uni<SearchResult> findFirstByTitle(String title) {
        return postRepository
                .find("#Post.findByTitle", Parameters.with("title", title))
                .firstResult()
                .map(post -> (SearchResult) new Found(post))
                .onFailure().recoverWithItem(error -> {
                    Log.error("Failed to find a post with title " + title, error);
                    return new NotFound(error);
                });
    }
}

File: src/main/kotlin/de/tungbeier/panache/PostService.kt
package de.tungbeier.panache

/*
import io.quarkus.hibernate.reactive.panache.PanacheRepository
import io.quarkus.logging.Log
import io.quarkus.panache.common.Parameters
import io.smallrye.mutiny.Uni
import jakarta.enterprise.context.ApplicationScoped
import org.hibernate.exception.ConstraintViolationException

@ApplicationScoped
class PostRepository : PanacheRepository<Post>

@ApplicationScoped
class PostService(private val repo: PostRepository) {

    fun register(post: Post): Uni<RegisterResult> = repo.persistAndFlush(post)
        .map {
            @Suppress("useless_cast")
            RegisterResult.Success(post) as RegisterResult
        }
        .onFailure().recoverWithItem { error ->
            Log.error("Failed to register a post", error)
            when (error) {
                is ConstraintViolationException -> RegisterResult.Conflict(post, error)
                else -> RegisterResult.Failure(post, error)
            }
        }

    fun findFirstPostByTitle(title: String): Uni<SearchResult> = repo
        .find("#Post.findByTitle", Parameters.with(Post::title.name, title))
        .firstResult<Post>()
        .map {
            @Suppress("useless_cast")
            SearchResult.Found(it) as SearchResult
        }
        .onFailure().recoverWithItem { error ->
            Log.error("Failed to search for post with title=$title", error)
            SearchResult.NotFound(error)
        }
}

sealed class RegisterResult(val post: Post) {
    class Success(post: Post) : RegisterResult(post)
    class Conflict(post: Post, val cause: Throwable) : RegisterResult(post)
    class Failure(post: Post, val cause: Throwable) : RegisterResult(post)
}

sealed class SearchResult {
    class Found(val post: Post) : SearchResult()
    class NotFound(val cause: Throwable) : SearchResult()
}
*/